Fix S3ToSnowflakeOperator to support uploading all files in the specified stage

sekikn opened the pull request:
Currently, users have to specify each file to upload as
the "s3_keys" parameter when using S3ToSnowflakeOperator.
But the `COPY INTO` statement, which S3ToSnowflakeOperator
leverages internally, allows omitting this parameter
so that users can upload whole files in the specified stage.
https://docs.snowflake.com/en/sql-reference/sql/copy-into-table.html#syntax

This PR makes S3ToSnowflakeOperator's s3_keys parameter optional
so as to support this functionality.

In addition to adding a corresponding test case, I manually ran the revised
version of S3ToSnowflakeOperator and confirmed that it worked, as follows:

----

Before executing S3ToSnowflakeOperator, no record is in the target table (via Snowflake CLI):

```
$ bin/snowsql --accountname *******.ap-southeast-1 --username sekikn --dbname FOO --schemaname BAR
Password: 
* SnowSQL * v1.2.10
Type SQL statements or !help
sekikn#COMPUTE_WH@FOO.BAR>LIST @S3_STG;
+-------------------------------------+------+----------------------------------+-------------------------------+
| name                                | size | md5                              | last_modified                 |
|-------------------------------------+------+----------------------------------+-------------------------------|
| s3://****/input/weather-snappy.avro |  330 | f1b4996b74a1a0de9e6ffa7d644f97a9 | Fri, 20 Nov 2020 02:42:15 GMT |
| s3://****/input/weather.avro        |  358 | 0a1cfdeca549207de10f431c8ed6dd4c | Fri, 20 Nov 2020 01:40:50 GMT |
+-------------------------------------+------+----------------------------------+-------------------------------+
2 Row(s) produced. Time Elapsed: 3.046s
sekikn#COMPUTE_WH@FOO.BAR>SELECT COUNT(*) FROM WEATHER;
+----------+                                                                    
| COUNT(*) |
|----------|
|        0 |
+----------+
1 Row(s) produced. Time Elapsed: 0.161s
```

Execute S3ToSnowflakeOperator without s3_keys (via Python REPL):

```
In [1]: from airflow.providers.snowflake.transfers.s3_to_snowflake import S3ToSnowflakeOperator
[2020-11-20 19:55:02,997] {arrow_result.pyx:0} INFO - Failed to import optional packages, pyarrow

In [2]: t = S3ToSnowflakeOperator(stage="FOO.BAR.S3_STG", file_format="(TYPE=AVRO)", schema="FOO.BAR", table="WEATHER", task_id="tid")

In [3]: t.execute(None)
[2020-11-20 19:55:13,968] {connection.py:210} INFO - Snowflake Connector for Python Version: 2.3.6, Python Version: 3.6.9, Platform: Linux-5.4.0-53-generic-x86_64-with-Ubuntu-18.04-bionic
[2020-11-20 19:55:13,968] {connection.py:744} INFO - This connection is in OCSP Fail Open Mode. TLS Certificates would be checked for validity and revocation status. Any other Certificate Revocation related exceptions or OCSP Responder failures would be disregarded in favor of connectivity.
[2020-11-20 19:55:13,968] {connection.py:760} INFO - Setting use_openssl_only mode to False
[2020-11-20 19:55:14,780] {cursor.py:531} INFO - query: [ALTER SESSION SET autocommit=True]
[2020-11-20 19:55:14,946] {cursor.py:553} INFO - query execution done
[2020-11-20 19:55:14,946] {dbapi_hook.py:179} INFO - Running statement: 
                COPY INTO FOO.BAR.WEATHER 
                    FROM @FOO.BAR.S3_STG/
                    
                    file_format=(TYPE=AVRO)
                
            , parameters: None
[2020-11-20 19:55:14,946] {cursor.py:531} INFO - query: [COPY INTO FOO.BAR.WEATHER FROM @FOO.BAR.S3_STG/  file_format=(TYPE=AVRO)]
[2020-11-20 19:55:19,840] {cursor.py:553} INFO - query execution done
[2020-11-20 19:55:19,840] {dbapi_hook.py:185} INFO - Rows affected: 2
[2020-11-20 19:55:19,840] {connection.py:430} INFO - closed
```

And all files are successfully uploaded (via Snowflake CLI):

```
sekikn#COMPUTE_WH@FOO.BAR>SELECT COUNT(*) FROM WEATHER;
+----------+                                                                    
| COUNT(*) |
|----------|
|       10 |
+----------+
1 Row(s) produced. Time Elapsed: 0.746s
```

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

sekikn committed:
Fix S3ToSnowflakeOperator to support uploading all files in the specified stage

Currently, users have to specify each file to upload as
the "s3_keys" parameter when using S3ToSnowflakeOperator.
But the `COPY INTO` statement, which S3ToSnowflakeOperator
leverages internally, allows omitting this parameter
so that users can upload whole files in the specified stage.
https://docs.snowflake.com/en/sql-reference/sql/copy-into-table.html#syntax

This PR makes S3ToSnowflakeOperator's s3_keys parameter optional
so as to support this functionality.

sekikn committed:
Fix static check errors

sekikn committed:
Fix pylint error

michalslowikowski00 commented on tests/providers/snowflake/transfers/test_s3_to_snowflake.py:
```diff
@@ -26,6 +26,41 @@
+            schema=schema, table=table, base_sql=base_sql
+        )
+
+        assert mock_run.call_count == 1
```
You can also use `mock_run.assert_called_once()`.

sekikn replied:
Thank you for the comment, @michalslowikowski00! Updated the PR.

michalslowikowski00 replied:
:)

sekikn committed:
Replace call_count with assert_called_once for code simplicity

dstandish commented on airflow/providers/snowflake/transfers/s3_to_snowflake.py:
```diff
@@ -71,16 +71,14 @@ def __init__(
 
         # we can extend this based on stage
         base_sql = """
                     FROM @{stage}/
```
Yeah so in addition to loading every file in stage, another pattern supported is having one stage per bucket and specify prefix e.g. `FROM '@{stage}/my-prefix/my-subprefix/v1/`

i think this PR would be a good opportunity to add support for that too.  perhaps it's as simple as adding a param `prefix` and making `from_path=f"{stage}/{prefix}" what do you think @sekikn?

though if you want to call that out of scope i'll withdraw the suggestion.

sekikn replied:
Thanks! My first thought was that users can specify prefix within the "stage" variable, but your design is cleaner actually. Just added it.
I manually confirmed that parameter worked, as follows:

Given an example S3 bucket with the following structure (each file contains five records),

```
$ aws s3 ls --recursive s3://****/input
2020-11-20 10:40:25          0 input/
2021-01-10 23:04:49          0 input/subdir/
2021-01-10 23:39:35        335 input/subdir/weather-sorted.avro
2020-11-20 11:42:15        330 input/weather-snappy.avro
2020-11-20 10:40:50        358 input/weather.avro
```

and the following stage and target table:

```
sekikn#COMPUTE_WH@FOO.BAR>DESC STAGE S3_STG;
+--------------------+---------------------+---------------+------------------------------------------------+------------------+
| parent_property    | property            | property_type | property_value                                 | property_default |
|--------------------+---------------------+---------------+------------------------------------------------+------------------|

(snip)

| STAGE_COPY_OPTIONS | FORCE               | Boolean       | true                                           | false            |
| STAGE_LOCATION     | URL                 | String        | ["s3://****/input/"]                           |                  |
| STAGE_INTEGRATION  | STORAGE_INTEGRATION | String        | S3_INT                                         |                  |

(snip)

sekikn#COMPUTE_WH@FOO.BAR>DESC TABLE WEATHER;
+------+---------+--------+-------+---------+-------------+------------+-------+------------+---------+
| name | type    | kind   | null? | default | primary key | unique key | check | expression | comment |
|------+---------+--------+-------+---------+-------------+------------+-------+------------+---------|
| C    | VARIANT | COLUMN | Y     | NULL    | N           | N          | NULL  | NULL       | NULL    |
+------+---------+--------+-------+---------+-------------+------------+-------+------------+---------+
1 Row(s) produced. Time Elapsed: 0.568s
sekikn#COMPUTE_WH@FOO.BAR>SELECT COUNT(*) FROM WEATHER;
+----------+                                                                    
| COUNT(*) |
|----------|
|        0 |
+----------+
1 Row(s) produced. Time Elapsed: 0.575s
```

First, ran S3ToSnowflakeOperator without prefix,

```
In [1]: from airflow.providers.snowflake.transfers.s3_to_snowflake import S3ToSnowflakeOperator
/home/sekikn/venv/a/lib/python3.8/site-packages/snowflake/connector/options.py:78 UserWarning: You have an incompatible version of 'pyarrow' installed (2.0.0), please install a version that adheres to: 'pyarrow<0.18.0,>=0.17.0; extra == "pandas"'

In [2]: t = S3ToSnowflakeOperator(stage="FOO.BAR.S3_STG", file_format="(TYPE=AVRO)", schema="FOO.BAR", table="WEATHER", task_id="tid")

In [3]: t.execute(None)
[2021-01-10 23:46:01,029] {connection.py:206} INFO - Snowflake Connector for Python Version: 2.3.6, Python Version: 3.8.5, Platform: Linux-5.8.0-36-generic-x86_64-with-glibc2.29
[2021-01-10 23:46:01,030] {connection.py:743} INFO - This connection is in OCSP Fail Open Mode. TLS Certificates would be checked for validity and revocation status. Any other Certificate Revocation related exceptions or OCSP Responder failures would be disregarded in favor of connectivity.
[2021-01-10 23:46:01,030] {connection.py:759} INFO - Setting use_openssl_only mode to False
[2021-01-10 23:46:01,783] {cursor.py:530} INFO - query: [ALTER SESSION SET autocommit=True]
[2021-01-10 23:46:01,927] {cursor.py:553} INFO - query execution done
[2021-01-10 23:46:01,927] {dbapi.py:180} INFO - Running statement: 
COPY INTO FOO.BAR.WEATHER
FROM @FOO.BAR.S3_STG/

file_format=(TYPE=AVRO)

, parameters: None
[2021-01-10 23:46:01,928] {cursor.py:530} INFO - query: [COPY INTO FOO.BAR.WEATHER FROM @FOO.BAR.S3_STG/  file_format=(TYPE=AVRO)]
[2021-01-10 23:46:02,657] {cursor.py:553} INFO - query execution done
[2021-01-10 23:46:02,658] {dbapi.py:186} INFO - Rows affected: 3
[2021-01-10 23:46:02,659] {connection.py:430} INFO - closed
```

and made sure 15 records were correctly uploaded via snowsql:

```
sekikn#COMPUTE_WH@FOO.BAR>SELECT COUNT(*) FROM WEATHER;
+----------+                                                                    
| COUNT(*) |
|----------|
|       15 |
+----------+
1 Row(s) produced. Time Elapsed: 1.126s
```

Then, ran S3ToSnowflakeOperator again **with** prefix:

```
In [4]: t = S3ToSnowflakeOperator(stage="FOO.BAR.S3_STG", prefix="subdir", file_format="(TYPE=AVRO)", schema="FOO.BAR", table="WEATHER", task_id="tid")

In [5]: t.execute(None)
[2021-01-10 23:47:56,478] {connection.py:206} INFO - Snowflake Connector for Python Version: 2.3.6, Python Version: 3.8.5, Platform: Linux-5.8.0-36-generic-x86_64-with-glibc2.29
[2021-01-10 23:47:56,478] {connection.py:743} INFO - This connection is in OCSP Fail Open Mode. TLS Certificates would be checked for validity and revocation status. Any other Certificate Revocation related exceptions or OCSP Responder failures would be disregarded in favor of connectivity.
[2021-01-10 23:47:57,155] {cursor.py:530} INFO - query: [ALTER SESSION SET autocommit=True]
[2021-01-10 23:47:57,290] {cursor.py:553} INFO - query execution done
[2021-01-10 23:47:57,290] {dbapi.py:180} INFO - Running statement: 
COPY INTO FOO.BAR.WEATHER
FROM @FOO.BAR.S3_STG/subdir

file_format=(TYPE=AVRO)

, parameters: None
[2021-01-10 23:47:57,291] {cursor.py:530} INFO - query: [COPY INTO FOO.BAR.WEATHER FROM @FOO.BAR.S3_STG/subdir  file_format=(TYPE=AVRO)]
[2021-01-10 23:47:57,947] {cursor.py:553} INFO - query execution done
[2021-01-10 23:47:57,948] {dbapi.py:186} INFO - Rows affected: 1
[2021-01-10 23:47:57,948] {connection.py:430} INFO - closed
```

and made sure only five records were additionally uploaded this time:

```
sekikn#COMPUTE_WH@FOO.BAR>SELECT COUNT(*) FROM WEATHER;
+----------+                                                                    
| COUNT(*) |
|----------|
|       20 |
+----------+
1 Row(s) produced. Time Elapsed: 0.628s
```

dstandish commented on tests/providers/snowflake/transfers/test_s3_to_snowflake.py:
```diff
@@ -26,6 +26,42 @@
+        assert_equal_ignore_multiple_spaces(self, mock_run.call_args[0][0], copy_query)
+
+    @mock.patch("airflow.providers.snowflake.hooks.snowflake.SnowflakeHook.run")
+    def test_execute_with_s3_keys(self, mock_run):
```
```suggestion
class TestS3ToSnowflakeTransfer(unittest.TestCase):
    @parameterized.expand([
        ('base', dict(table='table',
                      stage='stage',
                      file_format='file_format',
                      schema='schema',
                      ), "COPY INTO schema.table FROM @stage/ file_format=file_format"),
        ('files', dict(
            s3_keys=['1.csv', '2.csv'],
            table='table',
            stage='stage',
            file_format='file_format',
            schema='schema',
        ), "COPY INTO schema.table FROM @stage/ files=('1.csv', '2.csv') file_format=file_format"),
        ('columns', dict(
            table='table',
            stage='stage',
            file_format='file_format',
            schema='schema',
            columns_array=['col1', 'col2', 'col3'],
        ), "COPY INTO schema.table(col1,col2,col3) FROM @stage/ file_format=file_format"),
    ])
    @mock.patch("airflow.providers.snowflake.hooks.snowflake.SnowflakeHook.run")
    def test_execute(self, name, kwargs, expected, mock_run):
        S3ToSnowflakeOperator(
            **kwargs,
            task_id="task_id",
            dag=None,
        ).execute(None)

        mock_run.assert_called_once()
        assert_equal_ignore_multiple_spaces(self, mock_run.call_args[0][0], expected)
```
entire test can be replaced with this
using parameterization makes it much easier to read IMO

sekikn replied:
Agreed, update the test case to be parametrized.

sekikn committed:
Merge remote-tracking branch 'upstream/master' into s3_to_snowflake_without_files

sekikn committed:
Address @dstandish's comments

- Add prefix parameter to S3ToSnowflakeTransfer
- Update TestS3ToSnowflakeTransfer to be parameterized

sekikn committed:
Address static check error

sekikn committed:
Merge remote-tracking branch 'upstream/master' into s3_to_snowflake_without_files
